wallet: check for non-representable CFeeRates

jonatack opened the pull request:
As part of our migration to sat/vB feerates that began with #20305, this pull should resolve #20534 and improve/reduce the size of #20391 that adds a `setfeerate` RPC in sat/vB, as well as for a future `estimatefeerate` RPC in sat/vB. The code changes are less than 40 lines. Most of the diff is test coverage, as these changes touch a number of RPCs.

Changes:

- add `CFeeRate::FromSatB` and `CFeeRate::FromBtcKb` named constructors
- add a `CFeeRate::IsZero()` class member helper to replace `== CFeeRate(0)` conditionals and avoid object allocations/constructions
- add a `FeeRateFromValueInSatB()` utility function that calls `AmountFromValue()` and then performs an additional check for non-representable `CFeeRate` values in the range between 0 and 0.001 sat/vB
- add unit and functional tests and fill some current fee rate gaps in the functional test coverage

jonatack committed:
test: add non-representable CFeeRate test coverage

MarcoFalke commented:
the error message and error code should be the same when an invalid BTC/kb amount is rejected

MarcoFalke commented:
Concept ACK

jonatack commented:
> the error message and error code should be the same when an invalid BTC/kb amount is rejected

The other "too-low" messages I am aware of are:

Amount out of range
Fee rate (%s) is lower than the minimum fee rate setting
txfee cannot be less than min relay tx fee
txfee cannot be less than wallet min fee
txfee cannot be more than wallet max tx fee

The message proposed here seems more helpful, as it provides the minimum valid amount.

jonatack commented:
> the error message and error code should be the same when an invalid BTC/kb amount is rejected

The other "too-low" messages I am aware of are:

Amount out of range
Fee rate (%s) is lower than the minimum fee rate setting
txfee cannot be less than min relay tx fee
txfee cannot be less than wallet min fee

The message proposed here seems more helpful, as it provides the minimum valid amount.

MarcoFalke commented:
It would be confusing if 0.00000001 and 0.000000001 produced different error messages

jonatack commented:
fee_rate=0.000000001
```
fee_rate=0.0009999999
error code: -3
error message:
Invalid amount
```
...which is not a great error message, as it indicates neither which argument is invalid, nor why.

To stay close enough to not be negatively surprising, while providing better feedback, how about:

fee_rate=0.00000001
```
error code: -3
error message:
Invalid amount for fee_rate of 1e-08 (must be at least 0.001 sat/vB)
```

jonatack commented:
fee_rate=0.000000001
```
error code: -3
error message:
Invalid amount
```
...which is not a great error message, as it indicates neither which argument is invalid, nor why.

To stay close enough to not be negatively surprising, while providing better feedback, how about:

fee_rate=0.00000001
```
error code: -3
error message:
Invalid amount for fee_rate of 1e-08 (must be at least 0.001 sat/vB)
```

This is close to these similar error messages:

Invalid amount for -paytxfee=<amount>: '%s' (must be at least %s)
Invalid amount for -maxtxfee=<amount>: '%s' (must be at least the minrelay fee of %s to prevent stuck transactions)
Invalid amount for -fallbackfee=<amount>: '%s'
Invalid amount for -%s=<amount>: '%s

jonatack commented:
> Invalid amount
> ```
> 
> ...which is not a great error message, as it indicates neither which argument is invalid, nor why.

Maybe I should improve this.

jonatack committed:
test: add ParseFixedPoint fee rate rejection coverage

jonatack commented:
Just found a bug in RPC send. It only allows fee rates to be passed as a number, not as a string. Fixing with a commit here since it fits with the changes.

MarcoFalke commented:
> ...which is not a great error message, as it indicates neither which argument is invalid, nor why.

Then the error message should be fixed. But that is an issue that can be solve completely separate. It shouldn't be a reason to make the behaviour inconsistent.

jonatack committed:
test: add CFeeRate zero values functional test coverage

jonatack committed:
test: add type error and out of range fee rate coverage

jonatack commented:
> > ...which is not a great error message, as it indicates neither which argument is invalid, nor why.
> 
> Then the error message should be fixed. But that is an issue that can be solve completely separate. It shouldn't be a reason to make the behaviour inconsistent.

Done. Fixing the rpc send bug and improving the AmountFromValue error messages in separate pulls.

jonatack commented on test/functional/wallet_basic.py:
```diff
@@ -252,9 +252,15 @@ def run_test(self):
+            assert_raises_rpc_error(-3, msg, self.nodes[2].sendmany, amounts={address: 10}, fee_rate=invalid_value)
+        # Test values rejected by ParseFixedPoint() called in AmountFromValue()
+        for invalid_value in ["", 0.000000001, 1.111111111, 11111111111]:
+            assert_raises_rpc_error(-3, msg, self.nodes[2].sendmany, amounts={address: 1.0}, fee_rate=invalid_value)
```
Am testing CFeeRate errors separately from AmountFromValue errors, as a follow-up PR will improve the messages of both.

fjahr reviewed (commented):
TBH, I don't understand how this can close #20534 if the behavior is kept consistent. After merge the behavior described there is still the same. An error is caught at the "right" place internally but if the user can't tell the difference this rather seems to be a refactor?

jonatack commented:
The bug fix is that the response is consistent, though I agree it somewhat less user friendly. I have a follow-up that improves the error responses here and in AmountFromValue.

jonatack commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -765,8 +765,9 @@ def test_option_feerate(self):
                 node.fundrawtransaction, rawtx, {param: {"foo": "bar"}, "add_inputs": True})
-            assert_raises_rpc_error(-3, "Invalid amount",
-                node.fundrawtransaction, rawtx, {param: "", "add_inputs": True})
+            # Test values rejected by ParseFixedPoint() called in AmountFromValue()
```
We do refer to the implementation sometimes, perhaps most often when defining constants. 

This is to document why these values are tested separately for the same message. I have a follow-up to improve these messages in different ways, e.g. not just "Invalid amount" but also provide context to the user about which field and in some cases the valid range or values.

jonatack commented:
I wonder if this check shouldn't be in AmountFromValue and apply to fundrawtransaction and walletcreatefundedpsbt as well.￼￼

jonatack commented:
E.g. zero is ok but between zero and 0.001 sat/vB is invalid.

MarcoFalke commented:
the check should be in AmountFromValue. My issue is purely about parsing decimals, not about specific rpcs.

jonatack committed:
wallet: create FeeRateFromValueInSatB() util function

similar to AmountFromValue() with an additional check for non-representable
CFeeRates in the exclusive range of 0 to 0.001 sat/vB as CFeeRate rounds these
values down to CFeeRate(0)

jonatack committed:
policy: create CFeeRate::FromSatB/FromBtcKb named ctors

jonatack committed:
policy: create CFeeRate::IsZero() class member helper

and use it instead of constructing CFeeRates in conditionals

jonatack commented:
Pulled in the `CFeeRate::FromSatB` and `CFeeRate::FromBtcKb named constructors` from the setfeerate PR, added a `CFeeRate::IsZero()` class member helper, and used these to build a `FeeRateFromValueInSatB()` utility function. Added unit test coverage for each of these.

jonatack commented:
Pulled in the `CFeeRate::FromSatB` and `CFeeRate::FromBtcKb` named constructors from the setfeerate PR, added a `CFeeRate::IsZero()` class member helper, and used these to build a `FeeRateFromValueInSatB()` utility function. Added unit test coverage for each of these. This should handle the case of parsing decimals for fee rates constructed from sat/vB UniValue values.

jonatack commented on test/functional/wallet_send.py:
```diff
@@ -302,16 +302,31 @@ def run_test(self):
+        self.log.info("Explicit fee rate raises RPC error when invalid fee rates are passed")
+        # Test fee_rate with zero values
+        msg = "Fee rate (0.000 sat/vB) is lower than the minimum fee rate setting (1.000 sat/vB)"
+        for zero_value in [0, 0.000, 0.00000000]:
```
Note to self: update this line after #20573 is merged and `send` can take string fee rate values
```suggestion
        for zero_value in [0, 0.000, 0.00000000, "0", "0.000", "0.00000000"]:
```

jonatack replied:
updated

jonatack commented on test/functional/wallet_send.py:
```diff
@@ -302,16 +302,31 @@ def run_test(self):
+            self.test_send(from_wallet=w0, to_wallet=w1, amount=1, fee_rate=invalid_value, expect_error=(-3, msg))
+            self.test_send(from_wallet=w0, to_wallet=w1, amount=1, arg_fee_rate=invalid_value, expect_error=(-3, msg))
+        # Test fee_rate values rejected by ParseFixedPoint
+        for invalid_value in [0.000000001, 1.111111111, 11111111111]:
```
Note to self: update this line as well after #20573 is merged
```suggestion
        for invalid_value in ["", 0.000000001, 1.111111111, 11111111111]:
```

jonatack replied:
updated

jonatack commented:
rebased

jonatack commented:
Updated now that #20573 has been merged and rebased.

jonatack commented:
The ZMQ data race that causes the CI failure in https://cirrus-ci.com/task/6709748533821440?command=ci#L4578 is unrelated.

jonatack commented:
I really like these changes that do a lot of good things in ~40 lines and fix #20534, but seeing no enthusiasm here, closing.

jonatack commented:
I think these changes are (a) good on their own, (b) resolve issue #20534, and (c) unblock and reduce the size of #20391, so re-opening for review club to hopefully get some eyes here.

jonatack commented:
Suggest the following labels be added, please: `TX fees and policy`, `Refactoring`, `Review club`

jonatack commented:
Thanks @fanquake

felipsoarez reviewed (commented):
utACK

fanquake commented on src/test/rpc_tests.cpp:
```diff
@@ -214,6 +217,32 @@ BOOST_AUTO_TEST_CASE(rpc_parse_monetary_values)
+
+    // values rejected by ParseFixedPoint raise invalid amount
+    BOOST_CHECK_THROW(FeeRateFromValueInSatB(ValueFromString("0.000999999")), UniValue);
+    BOOST_CHECK_THROW(FeeRateFromValueInSatB(ValueFromString("0.000000001")), UniValue);
```
How far do we want to reach with these tests. This is basically "testing" that `FeeRateFromValueInSatB()` calls `AmountFromValue()` calls `ParseFixedPoint()` & `throw`s. Which is the same as:
```cpp
BOOST_CHECK_THROW(AmountFromValue(ValueFromString("0.000000001")), UniValue);
```
above, with an additional function call to get to `throw`.

However, because nothing specific is being checked, you could imagine how `FeeRateFromValueInSatB()` may end up being refactored such that there is some other throwing call, i.e:
```diff
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 1a83b9222..02604b595 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -89,6 +89,7 @@ CAmount AmountFromValue(const UniValue& value)
 
 CFeeRate FeeRateFromValueInSatB(const UniValue& value)
 {
+    this_might_throw();
     const CAmount amount{AmountFromValue(value)};
     const CFeeRate fee_rate{CFeeRate::FromSatB(amount)};
```

and the tests still pass, but only because `this_might_throw()` throws for some reason. Now what are we testing? This is also why I'm not a fan of the `raise invalid amount` type comments. That might be what we want the code to be testing, but that doesn't mean that's what it's doing.

jonatack replied:
Relooking at it in light of your thoughts here.

jonatack replied:
Dropped these unit tests and comments and combined the tests into the same commit as the function creation and the codebase updates.

fanquake reviewed (commented):
It would seem like we could do with some sort of fee rate specific functional test, as we are getting to the point where basically all wallet RPC related functional tests are having (essentially) the same blocks copied and pasted between all of them. This seems like an awful lot of duplication, where code easily get out of "sync", for not much additional coverage.

jonatack commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -768,6 +768,10 @@ def test_option_feerate(self):
                 node.fundrawtransaction, rawtx, {param: {"foo": "bar"}, "add_inputs": True})
             assert_raises_rpc_error(-3, "Invalid amount",
                 node.fundrawtransaction, rawtx, {param: "", "add_inputs": True})
+        # Test fee_rate values non-representable by CFeeRate
```
This is here for a reason. Fee rates should be accepted as an amount, e.g. a string or a number, as documented in the sendtoaddress, sendmany, fundrawtransaction, walletcreatefundedpsbt, send, and bumpfee helps. Adding this testing found a bug; see #20573.

fanquake replied:
If this is just to test that both amount and string fee rates are accepted, isn't that also tested by either
```python
for param, zero_value in product(["fee_rate", "feeRate"], [0, 0.000, 0.00000000, "0", "0.000", "0.00000000"]):
```
or
```python
for invalid_value in ["", 0.000000001, 1.111111111, 11111111111]:
```

(both added above) which use a mix of strings and numbers? Or the test changes made in #20573, which added specific tests for ["Test passing fee_rate as an integer"](https://github.com/bitcoin/bitcoin/blob/80486e7e2d8c360839cffc5a0c597b5c745433a7/test/functional/wallet_basic.py#L248) and ["Test passing fee_rate as a string"](https://github.com/bitcoin/bitcoin/blob/80486e7e2d8c360839cffc5a0c597b5c745433a7/test/functional/wallet_basic.py#L238)?

jonatack replied:
They are testing different things/codepaths, as documented (I'm grateful for the documentation in picking up the context again nearly two months later).

jonatack replied:
I reverified that:

- the "Test fee_rate values non-representable by CFeeRate" section verifies the exception raised in the new `FeeRateFromValueInSatB()` function (one can test this by changing the error message that function returns)
- the "Test fee_rate values rejected by ParseFixedPoint" section verifies the rejection by `ParseFixedPoint()` in `AmountFromValue()`
 
I'd like these cases to be tested separately to make sure the new function is raising on the amounts it is intended to. The next step is for both exceptions to raise better, more specific error messages.

jonatack commented on test/functional/wallet_basic.py:
```diff
@@ -273,6 +273,9 @@ def run_test(self):
         msg = "Invalid amount"
         for invalid_value in [0.00000001, 0.00099999, "0.00000001", "0.00099999"]:
             assert_raises_rpc_error(-3, msg, self.nodes[2].sendmany, amounts={address: 10}, fee_rate=invalid_value)
+        # Test fee_rate values rejected by ParseFixedPoint
```
My cited comment was in reply to a question about documentation. Which coverage should be dropped? It would have certainly been easier to not add coverage, but where fee rate test coverage has been partial or absent over the series of fee rate changes, I've ended up bitten by it and prefer to have the coverage in place for these changes and for future ones. The documentation is helpful to me when coming back to this months later (like now) to regain context. When I'm done with the fee rate work, I'm happy to remove the documentation.

jonatack replied:
Changed the comments in this commit to not name the specific function: "Test fee_rate values that don't pass fixed-point parsing checks".

jonatack commented:
> It would seem like we could do with some sort of fee rate specific functional test, as we are getting to the point where basically all wallet RPC related functional tests are having (essentially) the same blocks copied and pasted between all of them. This seems like an awful lot of duplication, where code easily get out of "sync", for not much additional coverage.

Well, only six (sendtoaddress, sendmany, send, fundrawtransaction, walletcreatefundedpsbt, and bumpfee) out of the sixty-some wallet/rpc functional tests, e.g. only for each of the send calls where it is important to have coverage.

Most of the work in #20220, #20305, #20426, and #20573 was writing tests, but they found a number of issues in different calls because coverage was previously partial or absent.

While a fee rate specific test is out of scope of this patch, I'm not sure it would reduce much duplication. Functional tests are end-to-end and need to exercise the endpoints or we'll end up back at square one with bugs. The coverage is unfortunately not copy-and-paste (I wish it were!), but having the test blocks look similar and cover evenly was a goal, because where the coverage was previously uneven, I came across issues, and it's pretty easy to omit coverage somewhere. The problem is that this is an finicky area where coverage helps and I was in the middle of a fair amount of changes. It's great to have it to protect against (some) future regressions and help future changes be easier and safer to make. At the same time, I do see that it's annoying to review and will look where some can be dropped or live on my computer only.

jonatack commented on src/policy/feerate.h:
```diff
@@ -34,6 +34,7 @@ class CFeeRate
 public:
     /** Fee rate of 0 satoshis per kB */
     CFeeRate() : nSatoshisPerK(0) { }
+    bool IsZero() const { return nSatoshisPerK == 0; }
```
Will do :+1:

jonatack replied:
Combined the `CFeeRate::IsZero()` creation, unit tests, and codebase updates into a single commit.

jonatack replied:
Idem for the named ctors and for the `FeeRateFromValueInSatB` utility function: creation, unit tests, and codebase updates all in a single commit.

theStack commented:
Concept ACK!
Unfortunately I missed the [review club for this PR](https://bitcoincore.reviews/20546) today, but went over the logs and it was a very interesting read.

Some thoughts:
* agree with [fanquake's suggestion](https://github.com/bitcoin/bitcoin/pull/20546/commits/d2b40010fab5a1ef7f6d946da076d4de6b442e09#r560606792) that commits https://github.com/bitcoin/bitcoin/pull/20546/commits/d2b40010fab5a1ef7f6d946da076d4de6b442e09 and https://github.com/bitcoin/bitcoin/pull/20546/commits/ed414f6dd72c103b5ba9e17c6b6bd2bcc8548b5b should be squashed; for newly introduced methods/functions that are really short, I barely see any benefit in splitting up the usage of it into another commit (other potential candidates are https://github.com/bitcoin/bitcoin/pull/20546/commits/ec62ce0997167b6110e029cc3b55afd97482fe4c + https://github.com/bitcoin/bitcoin/pull/20546/commits/ab4b12823dee3b8659ea2bc1fae43fb4efb6ecfb) generally for a merely 40 LOC change the number of commits seems quite large
* nice idea to introduce named constructors, makes the code way more readable 👍 
* with those, do we still need the two-parameter ctor `CFeeRate(const CAmount& nFeePaid, size_t nBytes)` to be public? all instances I see are either in test code or can be replaced by the named constructors now, so I guess there is not much point in exposing it

jonatack commented on src/test/rpc_tests.cpp:
```diff
@@ -214,6 +217,30 @@ BOOST_AUTO_TEST_CASE(rpc_parse_monetary_values)
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("93e+9")), UniValue); //overflow error
+
+    // values rejected by ParseFixedPoint raise invalid amount
+    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("0.000000001")), UniValue);
```
Dropped these assertions.

jonatack replied:
Dropped these assertions and comments.

jonatack commented:
> do we still need the two-parameter ctor `CFeeRate(const CAmount& nFeePaid, size_t nBytes)` to be public? all instances I see are either in test code or can be replaced by the named constructors now, so I guess there is not much point in exposing it

I looked at making it private but it is still used extensively in the codebase in critical code with the transaction size passed in as nBytes:
```
src/miner.cpp::241
src/node/psbt.cpp::140
src/policy/fees.cpp::549,579
src/txmempool.cpp::1037
src/validation.cpp::816,834
src/wallet/feebumper.cpp::86,120
```
(and in these tests)
```
src/test/amount_tests.cpp
src/test/mempool_tests.cpp
src/test/amount_tests.cpp
src/test/policyestimator_tests.cpp
```

jonatack commented:
Squashed a bunch of commits from 12 down to 7. Hopefully this is more pleasing to reviewers while remaining easy to review. In general, I try to make the changes look simple.

fjahr commented on src/test/amount_tests.cpp:
```diff
@@ -67,12 +72,14 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
 {
     // Test CFeeRate(CAmount fee_rate, size_t bytes) constructor
     // full constructor
+    BOOST_CHECK(CFeeRate(CAmount(0), 0).IsZero());
```
in b427ce178592283b9211b9fbd19874c7349bfd7e:

nit: I would have stuck these `IsZero()` tests in one place below since they are testing the function and not the ctor.

jonatack replied:
done

fjahr commented on src/test/amount_tests.cpp:
```diff
@@ -86,6 +90,31 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
+{
+    // Test CFeerate(CAmount fee_rate, FeeEstimatemode mode) constructor
+    // with BTC/kvB, returns same values as CFeeRate(amount) or CFeeRate(amount, 1000)
+    BOOST_CHECK(CFeeRate::FromBtcKb(CAmount(-1)) == CFeeRate(-1));
+    BOOST_CHECK(CFeeRate::FromBtcKb(CAmount(-1)) == CFeeRate(-1, 1000));
```
in afba3e188fcb14236324a0d1355445a99b85d155:

Aren't we already testing somewhere that `CFeeRate(-1) == CFeeRate(-1, 1000)`? Then I think we could save one of those and some other similar lines following.

jonatack replied:
Dropped that line and three other ones from this test.

fjahr reviewed (commented):
Concept ACK

Thanks for compressing this a bit, makes it easier to review for me at least. I had some thoughts but no major issues. Will follow up soon with another pass.

jonatack commented on src/policy/feerate.h:
```diff
@@ -70,4 +72,10 @@ class CFeeRate
 };
 
+/** Construct a CFeeRate from a CAmount in sat/vB */
+inline CFeeRate CFeeRate::FromSatB(CAmount fee_rate) { return CFeeRate(fee_rate, COIN); }
```
Here is a better explanation than I could provide: isocpp.org/wiki/faq/ctors#named-ctor-idiom by Stroustrup, Cline, Sutter and Alexandrescu

fjahr replied:
I did understand the idiom, the point I was trying to make that this should be functionally equivalent but just a bit simpler:

```
@ src/policy/feerate.h:43 @ public:
         // We've previously had bugs creep in from silent double->int conversion...
         static_assert(std::is_integral<I>::value, "CFeeRate should be used without floats");
     }
-    static CFeeRate FromSatB(CAmount fee_rate);
-    static CFeeRate FromBtcKb(CAmount fee_rate);
+
+    /** Construct a CFeeRate from a CAmount in sat/vB */
+    static CFeeRate FromSatB(CAmount fee_rate) { return CFeeRate(fee_rate, COIN); };
+    /** Construct a CFeeRate from a CAmount in BTC/kvB */
+    static CFeeRate FromBtcKb(CAmount fee_rate) { return CFeeRate(fee_rate, 1000); };
+
     /** Constructor for a fee rate in satoshis per kvB (sat/kvB). The size in bytes must not exceed (2^63 - 1).
      *
      *  Passing an nBytes value of COIN (1e8) returns a fee rate in satoshis per vB (sat/vB),
@ src/policy/feerate.h:80 @ public:
     SERIALIZE_METHODS(CFeeRate, obj) { READWRITE(obj.nSatoshisPerK); }
 };

-/** Construct a CFeeRate from a CAmount in sat/vB */
-inline CFeeRate CFeeRate::FromSatB(CAmount fee_rate) { return CFeeRate(fee_rate, COIN); }
-
-/** Construct a CFeeRate from a CAmount in BTC/kvB */
-inline CFeeRate CFeeRate::FromBtcKb(CAmount fee_rate) { return CFeeRate(fee_rate, 1000); }
-
 #endif //  BITCOIN_POLICY_FEERATE_H
```

It seems the inline definition outside of the body is mostly interesting if you want to do that multiple times in different translation units.

jonatack replied:
Oh! I'll look into it.

MarcoFalke replied:
I also prefer what @fjahr suggested, assuming it compiles

jonatack replied:
Thanks @fjahr. Your suggestion works; will use it if named constructors are needed in the future. Currently trying a different approach to this issue.

jonatack commented:
Thanks for reviewing @fjahr, updated with your feedback per `git diff 1fc25a4 236f556`

fjahr commented:
Code review ACK 236f556708fb619167b72cca451d0048a9274646

jarolrod reviewed (commented):
ACK 236f556708fb619167b72cca451d0048a9274646

This PR fixes the behavior shown in #20534

**Master:** Not representable, should through `error code: -3`
```
./src/bitcoin-cli -regtest -named sendtoaddress address=<addr> amount=1.23 fee_rate=0.0009

error code: -6
error message:
Fee rate (0.000 sat/vB) is lower than the minimum fee rate setting (1.000 sat/vB)
```

**PR:** throws `error code: -3`
```
./src/bitcoin-cli -regtest -named sendtoaddress address=<addr> amount=1.23 fee_rate=0.0009

error code: -3
error message:
Invalid amount
```

MarcoFalke commented:
(changed title because the only behaviour changes are in the wallet and this is not a refactor)

MarcoFalke commented on src/test/amount_tests.cpp:
```diff
@@ -111,6 +111,19 @@ BOOST_AUTO_TEST_CASE(CFeeRateNamedConstructorsTest)
+    BOOST_CHECK(CFeeRate(CAmount(1), 1001).IsZero());
+    BOOST_CHECK(CFeeRate::FromBtcKb(CAmount(0)).IsZero());
+    BOOST_CHECK(!CFeeRate::FromBtcKb(CAmount(1)).IsZero());
+    BOOST_CHECK(CFeeRate::FromSatB(CAmount(0.000)).IsZero());
```
in commit 83fb6101f0768e8f540e187caa2679c9f0b98c20: CAmount can't represent floats. This would be a compile error, if you didn't use a c-style-narrowing-cast

MarcoFalke replied:
For reference:

```
test/amount_tests.cpp:123:44: error: type 'double' cannot be narrowed to 'CAmount' (aka 'long') in initializer list [-Wc++11-narrowing]
    BOOST_CHECK(CFeeRate::FromSatB(CAmount{0.000}).IsZero());
                                           ^~~~~
```

jonatack replied:
changed to 0

MarcoFalke commented on src/rpc/util.cpp:
```diff
@@ -86,6 +87,17 @@ CAmount AmountFromValue(const UniValue& value)
+    const CAmount amount{AmountFromValue(value)};
+    const CFeeRate fee_rate{CFeeRate::FromSatB(amount)};
+    if (fee_rate.IsZero() && amount != 0) {
+        // passed value is an unrepresentable fee rate between 0 and 0.001 sat/vB
```
8885b724eaaba5ce610521d0ee00ce29a16310cc: Whether a feerate is representable or not has nothing to do with 0

9.99999999999999999999999999999 isn't representable and also isn't zero

MarcoFalke replied:
You need to fix this in `AmountFromValue`, as I pointed out earlier: https://github.com/bitcoin/bitcoin/pull/20546#issuecomment-739253473

jonatack replied:
Can you explain how and why this needs to be in `AmountFromValue`? Do we pass in an `is_sat_b_fee_rate` bool to it, so it has context? Is that an improvement over a dedicated function? I feel like I'm failing to read your mind on this since a couple months now.

jonatack replied:
dropped the comment

MarcoFalke replied:
`AmountFromValue` parses the string and fails if the amount is not representable, thus any check about representability should be in `AmountFromValue`.

btc/kb are representable when there are at most 8 digits. sat/b are representable when there are at most 3 digits. How you pass in the digits shouldn't matter. Can be done as normal argument or template argument (and function alias, if this makes the code easier to read).

maflcko reviewed (changes requested):
Approach re-NACK

Concept ACK

NACK 236f556708fb619167b72cca451d0048a9274646 😰

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

NACK 236f556708fb619167b72cca451d0048a9274646 😰
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgD5Av/Xab0IOByGeikbyUJLvFeyqNKCpgDZOwPp8IH3gMOw0RsQSOTba1Mw/mi
TYwJbgZs861puVMSvVGcdp3knH4GsE2SnhX/h3h3ozL1lEu0s4K0uksel7PtrwWH
Om4oCDT3kNNAcrGE/HBvCk723gx3JwT5zOWbBv6X//ESGrkUM9+yMbg3od4HBhmG
o24Ju8mugPXPKY9eQ+IG2RzXWItZPVVJIQBOVgD5awIBy15rgK50oOnLbT/8Mw8/
xbgaOZJzmt+3VczxsOTX+Qp+FMvX1fEJgO4Blsrk5bx2COy+tMst3lU16Vw9vcjh
huXxo7VdqiFPZB/q0BRg1Ysxl2o8o6ldKhrFm/W6Vtr5v3iIYxg7PTVjZAsJY98a
6DFpasP/KkWRwJSbqT63GETinOE+hKlu8sZO2X8R0NRGsyF+d9Lkyjq0FJsn4GDu
fcPe1KUYGa0nKQB7TvbZSBHxZ/eM84Qg1tCtX88rVLgXtuQCnFt92aW0l5PScw1I
4tC98cjZ
=4KcO
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `d9e37613c2875ff346833a9cd2c564a076e87e0eb46dd8c7fa743c1e891d0796  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108d9e37613c2875ff346833a9cd2c564a076e87e0eb46dd8c7fa743c1e891d0796f0102787efb1e6133ab41dd01242791144a708fff010c2b7fc7bec36290bd8d4b9a28f26aef608f10460111edaf008e47a00a89781de410083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010db07e4ec32b61f25b099444475dc105208f10460111edaf0081114187d0ce266280083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010ec741c5c7ac7694f57edc94c47529d3b08f10460111edaf008e5612e63ca802b280083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ed21001c100f3f5d15f9f9ed5ebcc17008f10460111edaf00802c64636e4b45e870083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

jonatack commented:
> This PR fixes the behavior shown in #20534
> 
> **Master:** Not representable, should through `error code: -3`
> 
> ```
> ./src/bitcoin-cli -regtest -named sendtoaddress address=<addr> amount=1.23 fee_rate=0.0009
> 
> error code: -6
> error message:
> Fee rate (0.000 sat/vB) is lower than the minimum fee rate setting (1.000 sat/vB)
> ```
> 
> **PR:** throws `error code: -3`
> 
> ```
> ./src/bitcoin-cli -regtest -named sendtoaddress address=<addr> amount=1.23 fee_rate=0.0009
> 
> error code: -3
> error message:
> Invalid amount
> ```

Thanks for testing! This shows that the error message here isn't as helpful as it could be, but I have a branch that improves the various AmountFromValue messages and this one as a follow-up.

jonatack commented:
> Approach re-NACK

Can you explain?  I've been failing to understand what is wrong with the approach for two months now, and I'm clearly not a very good mind reader.

jonatack commented:
> NACK 236f556

Can you explain why?

MarcoFalke commented:
> > NACK [236f556](https://github.com/bitcoin/bitcoin/commit/236f556708fb619167b72cca451d0048a9274646)
> 
> Can you explain why?

because it doesn't fix the bug. See:

```
$ ./src/bitcoin-cli -named sendtoaddress address=bcrt1qwkmhqum095zau5rf2velq55cq938vrd7uw8m4r amount=1.23 fee_rate=31.999999999999999999999
error code: -3
error message:
Invalid amount
$ ./src/bitcoin-cli -named sendtoaddress address=bcrt1qwkmhqum095zau5rf2velq55cq938vrd7uw8m4r amount=1.23 fee_rate=31.99999999
ac1ee7d06c8a276a148bf2378539c957bba643ba343f102779d618d5c95b9049
```

Both calls should fail because the amount is not representable

jonatack commented:
Ah, thanks for explaining. I thought that comment was referring to the commit.

jonatack commented:
> > Can you explain why?
> 
> because it doesn't fix the bug. See:
> ```
> $ ./src/bitcoin-cli -named sendtoaddress address=bcrt1qwkmhqum095zau5rf2velq55cq938vrd7uw8m4r amount=1.23 fee_rate=31.99999999
> ac1ee7d06c8a276a148bf2378539c957bba643ba343f102779d618d5c95b9049
> ```

This patch fixes the issue described in #20534 "values smaller than 0.001 sat/B can't be represented by CFeeRate, but they are not rejected" but I agree this should also be fixed. The approach here won't suffice so am trying a different one.

jonatack commented:
Closing for now in favor of #21786.
